File: codes/HellDream/278.java
public class Solution extends VersionControl {
    public int firstBadVersion(int n) {
        int start = 1, end = n;
        while(start+1<n){
            int mid = start+(n-start)/2;
            boolean isMidBad = isBadVersion(mid);
            // boolean isNBad = isBadVersion(n);
            boolean isStartBad = isBadVersion(start);
            if(!isMidBad&&!isStartBad) start = mid;
            else n = mid;

        }
        if(isBadVersion(start)) return start;
        if(isBadVersion(n)) return n;
        return 0;
    }
}

File: codes/Garnetwzy/278.cpp
// Forward declaration of isBadVersion API.
bool isBadVersion(int version);

class Solution {
public:
    int firstBadVersion(int n) {
        int s = 1, e = n;
        while(s < e) {
            int mid = s + (e - s) / 2;
            if(isBadVersion(mid)) {
                e = mid;
            }else {
                s = mid+1;
            }
        }
        return s;
    }
};